src/result_processor.c

```diff
@@ -845,3 +846,203 @@ ResultProcessor *RPCounter_New() {
+  ret->base.type = RP_BUFFER_AND_LOCKER;
+  return &ret->base;
+}
+
+static bool isResultValid(const SearchResult *res) {
+  // check if the doc is not marked deleted in the spec
+  return res->dmd->flags & Document_Deleted;
+}
+
+static int ReturnResult(SearchResult *buffered_result,  SearchResult *result_output) {
+  *result_output = *buffered_result;
+
+  // Invalidate the buffered result.
+  buffered_result->indexResult = NULL;
+  buffered_result->dmd = NULL;
+  buffered_result->scoreExplain = NULL;
+  buffered_result->indexResult = NULL;
+  memset(&buffered_result->rowdata, 0, sizeof(RLookupRow));
+  return RS_RESULT_OK;
+}
+
+static int rpbufferNext_Yield(ResultProcessor *rp, SearchResult *result_output) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+  SearchResult *curr_res = FixedSizeBlocksManager_getNextElement(&RPBuffer->resultsIterator);
+  
+  if(!curr_res) {
+    return RS_RESULT_EOF;
+  }
+ return ReturnResult(curr_res, result_output);
+  
+}
+
+static int rpbufferNext_ValidateAndYield(ResultProcessor *rp, SearchResult *result_output) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+  SearchResult *curr_res = FixedSizeBlocksManager_getNextElement(&RPBuffer->resultsIterator);
+  
+  // iterate the buffer.
+  while(curr_res) {
+    // Skip invalid results
+    if (isResultValid(curr_res)) {
+      return ReturnResult(curr_res, result_output);
+    }
+
+    // If the result is invalid discard it.
+    SearchResult_Clear(curr_res);
+    curr_res = FixedSizeBlocksManager_getNextElement(&RPBuffer->resultsIterator);
+  }
+
+  return RS_RESULT_EOF;
+}
+
+int rpbufferNext_bufferDocs(ResultProcessor *rp, SearchResult *res) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+
+  // Get the current index version so we can check later if any updates accured
+  // after we finish to buffer the results
+  RedisSearchCtx *RSctx = RPBuffer->base.parent->sctx;
+  double currentIndexVersion = IndexSpec_GetVersion(RSctx->spec);
+
+  FixedSizeBlocksManager *buffer = &RPBuffer->bufferBlocks;
+
+  // Allocate the first results block.
+  FixedSizeBlocksManager_init(buffer, sizeof(SearchResult),
+                                      RPBuffer->BlockSize);
+
+  // Keep fetching results from the upstream result processor until EOF is reached
+  while (1) {
+
+    // Get a space for a new result.
+    SearchResult *resToBuffer = FixedSizeBlocksManager_getEmptyElement(buffer);
+
+    // Get the next result and save it in the buffer
+    int result_status = rp->upstream->Next(rp->upstream, resToBuffer);
+
+    // If upstream has finished stop buffering results or something went wrong
+    if (result_status != RS_RESULT_OK) {
+      if (result_status == RS_RESULT_EOF ||
+          (result_status == RS_RESULT_TIMEDOUT &&
+           RSGlobalConfig.timeoutPolicy == TimeoutPolicy_Return)) {
+        // Break the loop and continue to the next step.
+        if (!FixedSizeBlocksManager_isEmpty(buffer)) {
```

Comment:
why do we check this condition here?

Reply:
optimization to return immediately without locking the GIL if there are zero results